index.js

```diff
@@ -125,6 +152,168 @@
+    return t.type == 'UNARY' || t.type == 'BINARY';
+  }
+  
+  var sigArgument = function(t, offset, isChild) {
+    var item = sigItem(t.type, offset,
+                       t.name ? stripNamespace(t.name) : t.toString(), ' ');
+    if (t.type == 'VARIABLE') {
+      item.typeVars[t.name] = sigTypeVar(t.name, offset);
+    } else if (isContainerType(t)) {
+      if (isChild) {
+        item.text = '(' + item.text;
+        item.end += 1;
+      }
+      var childTypes = [t.$1];
+      if (t.type == 'BINARY') {
+        childTypes.push(t.$2);
+      }
+      var idx, child;
+      for (idx = 0; idx < childTypes.length; idx += 1) {
+        item.end += item.glue.length;
+        child = sigArgument(childTypes[idx], item.end, true);
+        if (isContainerType(child)) {
+          item.children.push(child);
+        } else {
+          item.typeVars = merge(item.typeVars)(child.typeVars);
+        }
+        item.text += item.glue + child.text;
+        item.end = offset + item.text.length;
+      }
+      if (isChild) {
+        item.text += ')';
+        item.end += 1;
+      }
+    }
+    return item;
+  };
+  
+  var sigReturnType = function(expRetType, offset) {
+    var group = sigGroup(offset, '', '');
+    var item = sigArgument(expRetType, offset, false);
+    group.end = item.end;
+    group.items.push(item);
+    return group;
+  };
+  
+  var strRepeat = function(s, times) {
+    return Array(times + 1).join(s);
+  };
+  
+  var sigArgumentFindTypeVar = function(name) {
+    return function recur(item) {
+        var positions = name in item.typeVars ? [item.typeVars[name].offset] :
+                                                [];
+        return positions.concat(chain(item.children, recur));
+    };
+  };
+  
+  var sigShowTypeVarPositions = function(sig, name) {
+    var idx;
+    var positions = chain(sig.arguments.items, sigArgumentFindTypeVar(name));
+    
+    var upcarrots = '';
+    for (idx = 0; idx < positions.length; idx += 1) {
+      var pos = positions[idx];
+      upcarrots += strRepeat(' ', pos - upcarrots.length) + '^';
+    }
+    var varNumbers = '';
+    for (idx = 0; idx < positions.length; idx += 1) {
+      var pos = positions[idx];
+      varNumbers += strRepeat(' ', pos - varNumbers.length) + (idx + 1);
+    }
+    return upcarrots + '\n' + varNumbers;
+  };
+  
+  //  signature :: (String -> [{String: [TypeClass]}] -> [Type] -> Type) -> Object
+  var signature = function(name, constraints, expTypes, expRetType) {
+    var sig = {};
+    sig.name = sigName(name, 0);
+    sig.constraints = sigConstraints(constraints, sig.name.end);
+    sig.arguments = sigArguments(expTypes, sig.constraints.end);
+    sig.returnType = sigReturnType(expRetType, sig.arguments.end);
+    sig.toString = function() {
+      return (
+        sigGroupStr(sig.name) +
+        sigGroupStr(sig.constraints) +
+        sigGroupStr(sig.arguments) +
+        sigGroupStr(sig.returnType)
+      );
+    };
+    return sig;
+  };
```

Comment:
I've made some improvements to the code I posted above. Incorporate into your patch as much or as little of this code as your deem appropriate, @kedashoe. :)